### PR TITLE
feat: New Module `avm-res-compute-virtualmachine`

### PR DESCRIPTION
**Pull request by rahalan**
Add CARML VM module to AVM

[avm.res.compute.virtual-machine](https://github.com/rahalan/bicep-registry-modules/actions/workflows/avm.res.compute.virtual-machine.yml)

### CODE REVIEW

**Review by eriqua — CHANGES_REQUESTED** [review]
It looks like a lot of unintended changes sneaked in :snake:  (removed files and updates to utilities).

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -28,7 +28,7 @@
 #/avm/res/compute/image/ @Azure/avm-res-compute-image-module-owners-bicep @Azure/avm-core-team-technical-bicep
 #/avm/res/compute/proximity-placement-group/ @Azure/avm-res-compute-proximityplacementgroup-module-owners-bicep @Azure/avm-core-team-technical-bicep
 /avm/res/compute/ssh-public-key/ @Azure/avm-res-compute-sshpublickey-module-owners-bicep @Azure/avm-core-team-technical-bicep
-#/avm/res/compute/virtual-machine/ @Azure/avm-res-compute-virtualmachine-module-owners-bicep @Azure/avm-core-team-technical-bicep
+/avm/res/compute/virtual-machine/ @Azure/avm-res-compute-virtualmachine-module-owners-bicep @Azure/avm-core-team-technical-bicep
```
Comment (eriqua) [inline .github/CODEOWNERS:31]: Please make sure the corresponding team is created in the Azure organization and to follow the steps provided by the following link to assign the required access: https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr20---category-contributionsupport---github-teams-only

Comment (rahalan) [reply .github/CODEOWNERS:31]: I just created the Teams

**Review by eriqua — CHANGES_REQUESTED** [review]
Please merge main back to branch and rerun the pipeline, the PSRule job error should be fixed
Please also have a look at failing PSRule rules in defaults and waf-aligned tests. In particular:
- Ref open PR in CARML to fix rules related to the AMA agent
- Ref new option AZURE_DEPLOYMENT_NONSENSITIVE_PARAMETER_NAMES in [PSRule config file](https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/ps-rule.yaml) to suppress sensitive parameter false positives

File: avm/res/compute/virtual-machine/main.bicep [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+param patchAssessmentMode string = 'ImageDefault'
+
+@description('Optional. Specifies the time zone of the virtual machine. e.g. \'Pacific Standard Time\'. Possible values can be `TimeZoneInfo.id` value from time zones returned by `TimeZoneInfo.GetSystemTimeZones`.')
+param timeZone string = ''
+
+@description('Optional. Specifies additional base-64 encoded XML formatted information that can be included in the Unattend.xml file, which is used by Windows Setup. - AdditionalUnattendContent object.')
+param additionalUnattendContent array = []
+
+@description('Optional. Specifies the Windows Remote Management listeners. This enables remote Windows PowerShell. - WinRMConfiguration object.')
+param winRM object = {}
```
Comment (coolhome) [inline avm/res/compute/virtual-machine/main.bicep:248]: This should be an array or remove the mapping below to pass param `winRM` to property `winRM`

CARML PR - https://github.com/Azure/ResourceModules/pull/4388

**Review by coolhome — CHANGES_REQUESTED** [review]

File: avm/res/compute/virtual-machine/extension/main.json [hunk 3]
```diff
@@ -0,0 +1,160 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "languageVersion": "2.0",
+  "contentVersion": "1.0.0.0",
+  "metadata": {
+    "_generator": {
+      "name": "bicep",
+      "version": "0.23.1.45101",
```
Comment (eriqua) [inline avm/res/compute/virtual-machine/extension/main.json:8]: I wouldn't have asked normally, but since bicep 0.24 is out and we're regenerating all readmes with that version as part of another PR, could you please rerun set-avmmodule with the latest bicep version?

File: .github/workflows/avm.res.compute.virtual-machine.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      workflowInput: ${{ steps.get-workflow-param.outputs.workflowInput }}
+      moduleTestFilePaths: ${{ steps.get-module-test-file-paths.outputs.moduleTestFilePaths }}
+      psRuleModuleTestFilePaths: ${{ steps.get-module-test-file-paths.outputs.psRuleModuleTestFilePaths }}
+      modulePath: "${{ env.modulePath }}"
+
+  ##############################
+  #   Call reusable workflow   #
+  ##############################
+  call-workflow-passing-data:
+    name: "Module"
```
Comment (eriqua) [inline .github/workflows/avm.res.compute.virtual-machine.yml:74]: As per change which just got in for all other workflows
```suggestion
    name: "Run"
```

File: avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'dep-${namePrefix}-compute.virtualMachines-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'cvmlindef'
```
Comment (eriqua) [inline avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep:18]: Please use `min` instead of `def` here and in all default tests. Reason is we have a suppression group for PSRule filtering some rules on default tests which works on `min` substring. Once changed you'll see for example the tag rule will not complain anymore.
```suggestion
param serviceShort string = 'cvmlinmin'
```

Comment (AlexanderSehr) [reply avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep:18]: @eriqua this is actually a curios one. I guess the static tests needs updating, right? One would think this test should have caught it:

https://github.com/Azure/bicep-registry-modules/blob/c6850984b2f57645cddcd5cbc6aa84ca2e3faeff/avm/utilities/pipelines/staticValidation/compliance/module.tests.ps1#L1185-L1196

The issue seems to be that the test assumes a name like `[\/|\\]defaults\/|[\\]` eithen though it should be `[\/|\\].*defaults[\/|\\].+`

File: avm/res/compute/virtual-machine/extension/main.json (hunk 3 above)
Comment (rahalan) [reply avm/res/compute/virtual-machine/extension/main.json:8]: done

File: avm/res/compute/virtual-machine/main.bicep (hunk 2 above)
Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep]: done

File: avm/res/compute/virtual-machine/main.bicep [hunk 6]
```diff
@@ -0,0 +1,767 @@
...
+@allowed([
+  'AutomaticByPlatform'
+  'ImageDefault'
+])
+param patchAssessmentMode string = 'ImageDefault'
+
+@description('Optional. Specifies the time zone of the virtual machine. e.g. \'Pacific Standard Time\'. Possible values can be `TimeZoneInfo.id` value from time zones returned by `TimeZoneInfo.GetSystemTimeZones`.')
+param timeZone string = ''
+
+@description('Optional. Specifies additional base-64 encoded XML formatted information that can be included in the Unattend.xml file, which is used by Windows Setup. - AdditionalUnattendContent object.')
```
Comment (coolhome) [inline avm/res/compute/virtual-machine/main.bicep:241]: The following comes from the property description on `windowsConfiguration.additionalUnattendContent`:
```
Specifies additional base-64 encoded XML formatted information that can be included in the Unattend.xml file, which is used by Windows Setup.
```

Ref: https://learn.microsoft.com/en-us/azure/templates/microsoft.compute/virtualmachines?pivots=deployment-language-bicep#windowsconfiguration

The description for `additionalUnattendContent` type is better at describing this property in my opinion. Base64 is not relevant to the content passed to this parameter. My first instinct was to use `base64()` or `loadFileAsBase64()` which is rejected for unsupported format. 

```
Specifies additional XML formatted information that can be included in the Unattend.xml file, which is used by Windows Setup. Contents are defined by setting name, component name, and the pass in which the content is applied.
```

Ref: https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/create-or-update?view=rest-compute-2023-09-01&viewFallbackFrom=rest-compute-2023-07-01&tabs=HTTP#winrmconfiguration

**Review by coolhome — CHANGES_REQUESTED** [review]

File: avm/res/compute/virtual-machine/main.bicep (hunk 6 above)
Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep]: Thank you. Done.

File: avm/res/compute/virtual-machine/main.bicep [hunk 7]
```diff
@@ -0,0 +1,767 @@
...
+        telemetry: {
+          type: 'String'
+          value: 'For more information, see https://aka.ms/avm/TelemetryInfo'
+        }
+      }
+    }
+  }
+}
+
+module vm_nic 'br/public:avm/res/network/network-interface:0.2.2' = [for (nicConfiguration, index) in nicConfigurations: {
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/main.bicep:341]: Note: This module is missing the NIC+PIP deployment module from CARML: https://github.com/Azure/ResourceModules/blob/3e858dc2f82af7f5fe1864d803d127a077328d76/modules/compute/virtual-machine/main.bicep#L340

Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep:341]: @AlexanderSehr what should I add?

Comment (AlexanderSehr) [reply avm/res/compute/virtual-machine/main.bicep:341]: My ask would be to recover the [logic](https://github.com/Azure/ResourceModules/blob/main/modules/compute/virtual-machine/modules/nested_networkInterface.bicep) that was provided by the original CARML module which was capable of not just deploying n NICs but also n Public IPs for each NIC configuration. 
Was this removed on purpose? I know I pointed this change in design (i.e., removal) out before, but maybe there was a reason afterall.

Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep:341]: Will look into it. I guess I forgot. Sorry. Will keep you posted.

Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep:341]: @AlexanderSehr maybe I am not understanding, but the NIC is now a separate module that I created prior to the VM and it is used by the VM. What is missing?

Comment (AlexanderSehr) [reply avm/res/compute/virtual-machine/main.bicep:341]: Hey @rahalan that I'm not questioning and is exactly what you should do. But the original VM module never referenced the NIC module directly from the `main.bicep`, but instead an intermediate `nested_networkInterface.bicep` template that did a couple other things.
So my ask is to please recover that template and in it reference the external modules - just as we did in CARML (with the one difference that those were local file references).

Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep:341]: Hi @AlexanderSehr . After reviewing the code, I think you are asking for a change in the NIC module, not the VM one. Plus the NIC module already has the capability to deploy PIPs. Still not sure what needs to be changed.

File: avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            ]
+            subnetResourceId: nestedDependencies.outputs.subnetResourceId
+          }
+        ]
+        nicSuffix: '-nic-01'
+        tags: {
+          'hidden-title': 'This is visible in the resource name'
+          Environment: 'Non-Prod'
+          Role: 'DeploymentValidation'
+        }
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep:95]:
```suggestion
```

We should remove anything that isn't needed for this test to test / showcase it's purpose (auto-manage)

File: avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        ipConfigurations: [
+          {
+            name: 'ipconfig01'
+            pipConfiguration: {
+              publicIpNameSuffix: '-pip-01'
+              tags: {
+                'hidden-title': 'This is visible in the resource name'
+                Environment: 'Non-Prod'
+                Role: 'DeploymentValidation'
+              }
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep:80]:
```suggestion
```

We should remove anything that isn't needed for this test to test / showcase it's purpose (auto-manage)

File: avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+      {
+        keyData: nestedDependencies.outputs.SSHKeyPublicKey
+        path: '/home/localAdminUser/.ssh/authorized_keys'
+      }
+    ]
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Environment: 'Non-Prod'
+      Role: 'DeploymentValidation'
+    }
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep:118]:
```suggestion
```
We should remove anything that isn't needed for this test to test / showcase it's purpose (auto-manage)

File: avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  location: location
+}
+
+module nestedDependencies 'dependencies.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-nestedDependencies'
+  params: {
+    location: location
+    virtualNetworkName: 'dep-${namePrefix}-vnet-${serviceShort}'
+    managedIdentityName: 'dep-${namePrefix}-msi-${serviceShort}'
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep:40]: Unused dependency

File: avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+module nestedDependencies 'dependencies.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-nestedDependencies'
+  params: {
+    location: location
+    virtualNetworkName: 'dep-${namePrefix}-vnet-${serviceShort}'
+    sshDeploymentScriptName: 'dep-${namePrefix}-ds-${serviceShort}'
+    sshKeyName: 'dep-${namePrefix}-ssh-${serviceShort}'
+    managedIdentityName: 'dep-${namePrefix}-msi-${serviceShort}'
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep:42]: Unused dependency

File: avm/res/compute/virtual-machine/tests/e2e/waf-aligned/main.test.bicep [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+                eventHubName: diagnosticDependencies.outputs.eventHubNamespaceEventHubName
+                eventHubAuthorizationRuleResourceId: diagnosticDependencies.outputs.eventHubAuthorizationRuleId
+                storageAccountResourceId: diagnosticDependencies.outputs.storageAccountResourceId
+                workspaceResourceId: diagnosticDependencies.outputs.logAnalyticsWorkspaceResourceId
+              }
+            ]
+          }
+        ]
+        nicSuffix: '-nic-01'
+        roleAssignments: [
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/waf-aligned/main.test.bicep:137]: RBAC is not required to be WAF-aligned. Please also make sure to remove any other property that may be redundant. 
This may also mean that (like in this case) the MSI may be redundant.

File: avm/res/compute/virtual-machine/tests/e2e/windows.atmg/main.test.bicep [hunk 14]
```diff
@@ -0,0 +1,92 @@
+targetScope = 'subscription'
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/windows.atmg/main.test.bicep:1]: I think - to test/showcase atmg - it may be enough to have 1 test case, like the linux one. As such, I'd remove this test case.

File: avm/res/compute/virtual-machine/tests/e2e/windows.defaults/main.test.bicep [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+// ============== //
+// Test Execution //
+// ============== //
+@batchSize(1)
+module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
+  params: {
+    location: location
+    name: '${namePrefix}${serviceShort}'
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/windows.defaults/main.test.bicep:56]: Seems redundant for the default test

File: avm/res/compute/virtual-machine/tests/e2e/windows.ssecmk/main.test.bicep [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+            id: nestedDependencies.outputs.diskEncryptionSetResourceId
+          }
+        }
+      }
+    ]
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Environment: 'Non-Prod'
+      Role: 'DeploymentValidation'
+    }
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/tests/e2e/windows.ssecmk/main.test.bicep:108]:
```suggestion
```

May be redundant to show/test a deployment with a disk encryption set

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]

File: avm/res/compute/virtual-machine/main.bicep (hunk 7 above)
Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep]: @AlexanderSehr . Done. [avm.res.compute.virtual-machine](https://github.com/rahalan/bicep-registry-modules/actions/workflows/avm.res.compute.virtual-machine.yml)

File: avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+//   scope: resourceGroup
+// }
+
+@batchSize(1)
+module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
+  params: {
+    location: location
+    name: '${namePrefix}${serviceShort}'
```
[the remark this answers is not in the document]
Comment (rahalan) [reply avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep:60]: done

[the remark this answers is not in the document]
Comment (rahalan) [reply avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep:60]: rule BCP179 is not allowing that

File: avm/res/compute/virtual-machine/main.bicep [hunk 18]
```diff
@@ -0,0 +1,768 @@
...
+        telemetry: {
+          type: 'String'
+          value: 'For more information, see https://aka.ms/avm/TelemetryInfo'
+        }
+      }
+    }
+  }
+}
+
+module vm_nic 'modules/nic-configuration.bicep' = [for (nicConfiguration, index) in nicConfigurations: {
```
[the remark this answers is not in the document]
Comment (rahalan) [reply avm/res/compute/virtual-machine/main.bicep:341]: done

File: avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep (hunk 12 above)
Comment (rahalan) [reply avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep:42]: which one?

File: avm/res/compute/virtual-machine/tests/e2e/waf-aligned/main.test.bicep (hunk 13 above)
Comment (rahalan) [reply avm/res/compute/virtual-machine/tests/e2e/waf-aligned/main.test.bicep]: done

File: avm/res/compute/virtual-machine/tests/e2e/windows.atmg/main.test.bicep (hunk 14 above)
Comment (rahalan) [reply avm/res/compute/virtual-machine/tests/e2e/windows.atmg/main.test.bicep:1]: done

File: avm/res/compute/virtual-machine/tests/e2e/windows.defaults/main.test.bicep (hunk 15 above)
Comment (rahalan) [reply avm/res/compute/virtual-machine/tests/e2e/windows.defaults/main.test.bicep:56]: name is mandatory

File: avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep (hunk 17 above)
[the remark this answers is not in the document]
Comment (AlexanderSehr) [reply avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep:61]: In that case I'd recommend removing the dafault. I'm anyways not convinced it makes much sense to generate a VM name out the resource group name. Thoughts?

File: avm/res/compute/virtual-machine/modules/nic-configuration.bicep [hunk 19]
```diff
@@ -0,0 +1,152 @@
+param networkInterfaceName string
+param virtualMachineName string
+param location string
+param tags object?
+param enableIPForwarding bool = false
+param enableAcceleratedNetworking bool = false
+param dnsServers array = []
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableTelemetry bool = true
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/modules/nic-configuration.bicep:9]:
```suggestion

@description('Required. Enable telemetry via a Globally Unique Identifier (GUID).')
param enableTelemetry bool
```

Would recommend to make it recommended here so that it's not forgotten in the future.

File: avm/res/compute/virtual-machine/modules/nic-configuration.bicep [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+    roleAssignments: contains(ipConfiguration, 'roleAssignments') ? ipConfiguration.roleAssignments : []
+    skuName: contains(ipConfiguration, 'skuName') ? ipConfiguration.skuName : 'Standard'
+    skuTier: contains(ipConfiguration, 'skuTier') ? ipConfiguration.skuTier : 'Regional'
+    tags: ipConfiguration.?tags ?? tags
+    zones: contains(ipConfiguration, 'zones') ? ipConfiguration.zones : [
+      '1'
+      '2'
+      '3'
+    ]
+    enableTelemetry: enableTelemetry
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/modules/nic-configuration.bicep:42]: As per previous comment, should be:
```suggestion
    enableTelemetry: ipConfiguration.?enableTelemetry ?? enableTelemetry
```

File: avm/res/compute/virtual-machine/modules/nic-configuration.bicep [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+
+@description('Optional. The diagnostic settings of the Network Interface.')
+param diagnosticSettings diagnosticSettingType
+
+@description('Optional. Array of role assignments to create.')
+param roleAssignments roleAssignmentType
+
+var enableReferencedModulesTelemetry = false
+
+module networkInterface_publicIPAddresses 'br/public:avm/res/network/public-ip-address:0.2.1' = [for (ipConfiguration, index) in ipConfigurations: if (contains(ipConfiguration, 'pipconfiguration')) {
```
[the remark this answers is not in the document]
Comment (AlexanderSehr) [reply avm/res/compute/virtual-machine/modules/nic-configuration.bicep:23]: Resolving in favor of other comment

File: avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep (hunk 12 above)
Comment (AlexanderSehr) [reply avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep:42]: The `managedIdentity` that I'm commenting. No?

File: avm/res/compute/virtual-machine/tests/e2e/windows.defaults/main.test.bicep (hunk 15 above)
Comment (AlexanderSehr) [reply avm/res/compute/virtual-machine/tests/e2e/windows.defaults/main.test.bicep:56]: Consider this covered by my other reply. Resolving.

File: avm/res/compute/virtual-machine/main.bicep [hunk 22]
```diff
@@ -0,0 +1,769 @@
...
+  @description('Optional. The name of logs that will be streamed. "allLogs" includes all possible logs for the resource. Set to \'\' to disable log collection.')
+  logCategoriesAndGroups: {
+    @description('Optional. Name of a Diagnostic Log category for a resource type this setting is applied to. Set the specific logs to collect here.')
+    category: string?
+
+    @description('Optional. Name of a Diagnostic Log category group for a resource type this setting is applied to. Set to \'AllLogs\' to collect all logs.')
+    categoryGroup: string?
+  }[]?
+
+  @description('Optional. The name of logs that will be streamed. "allLogs" includes all possible logs for the resource. Set to \'\' to disable log collection.')
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/main.bicep:746]:
```suggestion
  @description('Optional. The name of metrics that will be streamed. "allMetrics" includes all possible metrics for the resource. Set to \'\' to disable metric collection.')
```

File: avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep (hunk 17 above)
[the remark this answers is not in the document]
Comment (rahalan) [reply avm/res/compute/virtual-machine/tests/e2e/linux.defaults/main.test.bicep:61]: Done

File: avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep (hunk 12 above)
Comment (rahalan) [reply avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep:42]: It is needed for the sshDeploymentScript

File: avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep [hunk 23]
```diff
@@ -0,0 +1,123 @@
+targetScope = 'subscription'
```
[the remark this answers is not in the document]
Comment (rahalan) [reply avm/res/compute/virtual-machine/tests/e2e/linux.atmg/main.test.bicep:1]: Done

File: avm/res/compute/virtual-machine/main.bicep [hunk 24]
```diff
@@ -353,7 +353,7 @@ module vm_nic 'modules/nic-configuration.bicep' = [for (nicConfiguration, index)
     tags: nicConfiguration.?tags ?? tags
     diagnosticSettings: nicConfiguration.?diagnosticSettings
     roleAssignments: nicConfiguration.?roleAssignments
-    enableTelemetry: nicConfiguration.?enableTelemetry ?? enableTelemetry
+    enableTelemetry: enableTelemetry
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/main.bicep:356]:
```suggestion
    enableTelemetry: enableTelemetry: nicConfiguration.?enableTelemetry ?? enableTelemetry
```

Why did you remove this?

File: .github/workflows/avm.res.compute.virtual-machine.yml [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+
+concurrency:
+  group: ${{ github.workflow }}
+
+jobs:
+  ###########################
+  #   Initialize pipeline   #
+  ###########################
+  job_initialize_pipeline:
+    runs-on: ubuntu-20.04
```
Comment (eriqua) [inline .github/workflows/avm.res.compute.virtual-machine.yml:47]: As per recent change to avm workflows
```suggestion
    runs-on: ubuntu-latest
```

File: avm/res/compute/virtual-machine/main.bicep [hunk 26]
```diff
@@ -0,0 +1,769 @@
...
+@description('Optional. The lock settings of the service.')
+param lock lockType
+
+@description('Optional. Array of role assignments to create.')
+param roleAssignments roleAssignmentType
+
+@description('Optional. Tags of the resource.')
+param tags object?
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/main.bicep:195]:
```suggestion
@description('Optional. Enable/Disable usage telemetry for module.')
```

In alignment with the specs. Requires a regeneration of the files

File: avm/res/compute/virtual-machine/main.bicep [hunk 27]
```diff
@@ -0,0 +1,769 @@
...
+// resource vm_logAnalyticsWorkspace 'Microsoft.OperationalInsights/workspaces@2021-06-01' existing = if (!empty(monitoringWorkspaceId)) {
+//   name: last(split((!empty(monitoringWorkspaceId) ? monitoringWorkspaceId : 'law'), '/'))!
+//   scope: az.resourceGroup(split((!empty(monitoringWorkspaceId) ? monitoringWorkspaceId : '//'), '/')[2], split((!empty(monitoringWorkspaceId) ? monitoringWorkspaceId : '////'), '/')[4])
+// }
+
+module vm_azureMonitoringAgentExtension 'extension/main.bicep' = if (extensionMonitoringAgentConfig.enabled) {
+  name: '${uniqueString(deployment().name, location)}-VM-AzureMonitoringAgent'
+  params: {
+    virtualMachineName: vm.name
+    name: 'AzureMonitoringAgent'
```
Comment (AlexanderSehr) [inline avm/res/compute/virtual-machine/main.bicep:527]:
```suggestion
    name: 'AzureMonitorAgent'
```

I'm not sure if the name matters? Just adding the suggestion here as I see the original proposal by @fabmas used a different name.

**Review by eriqua — APPROVED** [review]

**Comment by eriqua** [comment]
@rahalan this is good for me related to changes we discussed. I'll leave the last word on the module to @AlexanderSehr who requested changes there.

**Review by AlexanderSehr — APPROVED** [review]
